virtualizarr/manifests/store.py

```diff
@@ -0,0 +1,367 @@
+                store,
+                (
+                    obs.store.AzureStore,
+                    obs.store.GCSStore,
+                    obs.store.HTTPStore,
+                    obs.store.S3Store,
+                    obs.store.LocalStore,
+                    obs.store.MemoryStore,
+                ),
+            ):
+                raise TypeError(f"expected ObjectStore class, got {store!r}")
+        # TODO: Don't allow stores with prefix
+        # TODO: Type check the manifest arrays
+        super().__init__(read_only=True)
+        self._stores = stores
+        self._manifest_group = manifest_group
+
+    def __str__(self) -> str:
+        return f"ManifestStore({self._manifest_group}, {self._stores})"
+
+    def __getstate__(self) -> dict[Any, Any]:
+        state = self.__dict__.copy()
+        stores = state["_stores"].copy()
+        for k, v in stores.items():
+            stores[k] = pickle.dumps(v)
+        state["_stores"] = stores
+        return state
+
+    def __setstate__(self, state: dict[Any, Any]) -> None:
+        stores = state["_stores"].copy()
+        for k, v in stores.items():
+            stores[k] = pickle.loads(v)
+        state["_stores"] = stores
+        self.__dict__.update(state)
+
+    async def get(
+        self,
+        key: str,
+        prototype: BufferPrototype,
+        byte_range: ByteRequest | None = None,
+    ) -> Buffer | None:
+        # docstring inherited
+        import obstore as obs
+
+        if key.endswith("zarr.json"):
+            return get_zarr_metadata(self._manifest_group, key)
+        var, chunk_key = parse_manifest_index(key)
+        marr = self._manifest_group._manifest_dict[var]
+        manifest = marr._manifest
+
+        path = manifest._paths[*chunk_key]
+        offset = manifest._offsets[*chunk_key]
+        length = manifest._lengths[*chunk_key]
+        store_request = find_matching_store(stores=self._stores, request_key=path)
+        # Get the  configured object store instance that matches the path
+        store = self._stores[store_request.store_id]
+        # Transform the input byte range to account for the chunk location in the file
+        chunk_end_exclusive = offset + length
+        byte_range = _transform_byte_range(
+            byte_range, chunk_start=offset, chunk_end_exclusive=chunk_end_exclusive
+        )
+        # Actually get the bytes
+        try:
+            bytes = await obs.get_range_async(
+                store,
+                store_request.key,
+                start=byte_range.start,
+                end=byte_range.end,
+            )
+            return prototype.buffer.from_bytes(bytes)  # type: ignore[arg-type]
+        except _ALLOWED_EXCEPTIONS:
+            return None
+
+    async def get_partial_values(
+        self,
+        prototype: BufferPrototype,
+        key_ranges: Iterable[tuple[str, ByteRequest | None]],
+    ) -> list[Buffer | None]:
+        # docstring inherited
+        raise NotImplementedError
```

Comment:
If you plan to implement this in the future, you could even consider importing the private functions from the upstream zarr obstore file.

Reply:
thanks for the note, I added a TODO in [`709788b` (#490)](https://github.com/zarr-developers/VirtualiZarr/pull/490/commits/709788b0451f33d8552b0315c1896566ff06a00a)